xapian-applications/omega/assistant.cc

```diff
@@ -0,0 +1,99 @@
+/** @file assistant.cc
+ * @brief Worker module for putting text extraction into a separate process.
+ */
+/* Copyright (C) 2011 Olly Betts
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 2 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301
+ * USA
+ */
+
+#include <config.h>
+#include <csignal>
+
+#include "worker_comms.h"
+#include "handler.h"
+#include "safeunistd.h"
+
+using namespace std;
+
+const int FD = 3;
+const int time_limit = 300;
+
+#if defined HAVE_ALARM
+
+static void
+timeout_handler(int sig) {
+    _Exit(sig);
+}
+
+static void
+set_timeout() {
+    // Causes the system to generate a SIGALRM signal for the process after
+    // time_limit of seconds
+    alarm(time_limit);
+    // Set a signal handler for SIGALRM
+    signal(SIGALRM, timeout_handler);
+}
+
+static void
+stop_timeout() {
+    // Set SIGALRM to be ignore
+    signal(SIGALRM, SIG_IGN);
+    // A pending alarm request, if any, is cancelled by calling alarm(0)
+    alarm(0);
+}
+
+#else
+
+static void set_timeout() { }
+static void stop_timeout() { }
+
+#endif
+
+// FIXME: Restart filter every N files processed?
+
+int main() {
+
+    string filename;
+    FILE * sockt = fdopen(FD, "r+");
+
+    // Just in case (if the pid of this new process is equal
+    // to an old assistant pid it is possible to have problems)
+    stop_timeout();
```

Comment:
What's the potential problem here?

Reply:
It may be a very extreme case, but I was not sure what could happen if an assistant process started with the pid of an old assistant process that had ended due to an error (such as SEGFAULT). If this were the case, the new assistant could fail for the old timeout (Assuming that the operating system does not eliminate the signal when a process ends). So I cancel any request just in case.
However, I didn't realize that the alarm would be ignore as the handler has not been set yet.
This line of code is unnecessary. I am going to remove it in the next commit.

Reply:
I'd expect the operating system to clear such state when a process exits.

If it didn't, you'd see problems more widely as processes would get apparently random SIGALRM signals, and the default handling for SIGALRM is to terminate the process.

Good to be thinking about how things can go wrong though.